core/src/main/java/com/sequenceiq/cloudbreak/converter/stack/cluster/ServiceEndpointCollector.java

```diff
@@ -0,0 +1,189 @@
+            StackServiceComponentDescriptor componentDescriptor, Cluster cluster) {
+        Map<String, String> collectedPorts = new HashMap<>();
+        if (componentDescriptor != null && componentDescriptor.isMaster()) {
+            Gateway gateway = cluster.getGateway();
+            if (gateway != null) {
+                Set<String> exposedServices = gateway.getTopologies()
+                        .stream()
+                        .flatMap(this::getExposedServiceStream)
+                        .filter(Objects::nonNull)
+                        .collect(Collectors.toSet());
+
+                for (Port port : ports) {
+                    Map<String, String> result = collectServiceUrlsForPorts(port, serviceAddress, ambariIp, componentDescriptor, exposedServices, gateway);
+                    collectedPorts.putAll(result);
+                }
+            }
+        }
+        return collectedPorts;
+    }
+
+    private Stream<String> getExposedServiceStream(GatewayTopology gatewayTopology) {
+        if (gatewayTopology.getExposedServices() != null && gatewayTopology.getExposedServices().getValue() != null) {
+            try {
+                return gatewayTopology.getExposedServices().get(ExposedServices.class).getServices().stream();
+            } catch (IOException e) {
+                LOGGER.warn("Failed to get exposed services from Json.", e);
+            }
+        }
+        return Stream.empty();
+    }
+
+    private Map<String, String> collectServiceUrlsForPorts(Port port, String serviceAddress, String ambariIp,
+            StackServiceComponentDescriptor componentDescriptor, Collection<String> exposedServices, Gateway gateway) {
+        Map<String, String> collectedPorts = new HashMap<>();
+        if (port.getExposedService().getServiceName().equalsIgnoreCase(componentDescriptor.getName())) {
+            if (ambariIp != null) {
+                gateway.getTopologies()
+                        .stream()
+                        .map(gt -> getServiceUrlForPort(port, ambariIp, exposedServices, gateway, gt.getTopologyName()))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .forEach(url -> collectedPorts.put(port.getExposedService().getPortName(), url));
+            } else if (serviceAddress != null) {
+                String url = String.format("http://%s:%s%s", serviceAddress, port.getPort(), port.getExposedService().getPostFix());
+                collectedPorts.put(port.getExposedService().getPortName(), url);
+            }
+        }
+        return collectedPorts;
+    }
+
+    private Optional<String> getServiceUrlForPort(Port port, String ambariIp, Collection<String> exposedServices, Gateway gateway, String topologyName) {
+        String url = GatewayType.CENTRAL == gateway.getGatewayType() ? String.format("/%s/%s%s", gateway.getPath(), topologyName,
```

Comment:
Would be nicer to format expression by `?` and `:` instead of the middle of a long format